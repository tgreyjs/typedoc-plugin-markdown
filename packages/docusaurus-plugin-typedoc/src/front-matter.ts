import * as path from 'path';

import { BindOption } from 'typedoc';
import {
  getPageTitle,
  prependYAML,
} from 'typedoc-plugin-markdown/dist/utils/front-matter';
import { Component } from 'typedoc/dist/lib/converter/components';
import { RendererComponent } from 'typedoc/dist/lib/output/components';
import { PageEvent } from 'typedoc/dist/lib/output/events';

import { FrontMatter, Sidebar } from './types';

@Component({ name: 'front-matter' })
export class FrontMatterComponent extends RendererComponent {
  @BindOption('out')
  out!: string;
  @BindOption('sidebar')
  sidebar!: Sidebar;
  @BindOption('globalsTitle')
  globalsTitle!: string;
  @BindOption('readmeTitle')
  readmeTitle!: string;
  @BindOption('entryDocument')
  entryDocument!: string;
  @BindOption('entryPoints')
  entryPoints!: string[];

  globalsFile = 'modules.md';

  initialize() {
    super.initialize();
    this.listenTo(this.application.renderer, {
      [PageEvent.END]: this.onPageEnd,
    });
  }
  onPageEnd(page: PageEvent) {
    if (page.contents) {
      page.contents = prependYAML(page.contents, this.getYamlItems(page));
    }
  }

  getYamlItems(page: PageEvent): any {
    const pageId = this.getId(page);
    const pageTitle = this.getTitle(page);
    const sidebarLabel = this.getSidebarLabel(page);
    const sidebarPosition = this.getSidebarPosition(page);
    let items: FrontMatter = {
      id: pageId,
      title: pageTitle,
    };
    if (page.url === this.entryDocument) {
      items = { ...items, slug: '/' + this.out };
    }
    if (sidebarLabel && sidebarLabel !== pageTitle) {
      items = { ...items, sidebar_label: sidebarLabel };
    }
    if (sidebarPosition) {
      items = { ...items, sidebar_position: parseFloat(sidebarPosition) };
    }
    if (page.url === page.project.url && this.entryPoints.length > 1) {
      items = { ...items, hide_table_of_contents: true };
    }
    return {
      ...items,
      custom_edit_url: null,
    };
  }

  getSidebarLabel(page: PageEvent) {
    const indexLabel =
      this.sidebar.indexLabel ||
      (this.entryPoints.length > 1 ? 'Table of contents' : 'Exports');

    if (page.url === this.entryDocument) {
      return page.url === page.project.url
        ? indexLabel
        : this.sidebar.readmeLabel;
    }

    if (page.url === this.globalsFile) {
      return indexLabel;
    }

    return this.sidebar.fullNames ? page.model.getFullName() : page.model.name;
  }

  getSidebarPosition(page: PageEvent) {
    if (page.url === this.entryDocument) {
      return page.url === page.project.url ? '0.5' : '0';
    }
    if (page.url === this.globalsFile) {
      return '0.5';
    }
    if (page.model.getFullName().split('.').length === 1) {
      return '0';
    }
    return null;
  }

  getId(page: PageEvent) {
    return path.basename(page.url, path.extname(page.url));
  }

  getTitle(page: PageEvent) {
    const readmeTitle = this.readmeTitle || page.project.name;
    if (page.url === this.entryDocument && page.url !== page.project.url) {
      return readmeTitle;
    }
    return getPageTitle(page);
  }
}
