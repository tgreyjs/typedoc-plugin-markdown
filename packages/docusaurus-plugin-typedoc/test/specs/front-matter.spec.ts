import cuid from 'cuid';
import { Application } from 'typedoc';
import MarkdownPlugin from 'typedoc-plugin-markdown';
import { PageEvent } from 'typedoc/dist/lib/output/events';

import { FrontMatterComponent } from '../../dist/front-matter';
import { bootstrap } from '../../dist/render';
import { PluginOptions } from '../../dist/types';

async function generate(
  opts = {},
  entryPoints = [
    '../typedoc-plugin-markdown/test/stubs/src/theme.ts',
    '../typedoc-plugin-markdown/test/stubs/src/frontmatter.ts',
  ],
) {
  const app = new Application();

  MarkdownPlugin(app);

  bootstrap(app, {
    ...opts,
    entryPoints,
    tsconfig: '../typedoc-plugin-markdown/test/stubs/tsconfig.json',
  } as Partial<PluginOptions>);

  const project = app.convert();
  const componentNamename = cuid();
  app.renderer.addComponent(
    componentNamename,
    new FrontMatterComponent(app.renderer),
  );
  const frontMatterComponent = app.renderer.getComponent(
    componentNamename,
  ) as FrontMatterComponent;
  return { project, frontMatterComponent };
}

describe(`FrontMatter:`, () => {
  describe(`(readme)`, () => {
    test(`should set default index page`, async () => {
      const page = {
        url: 'index.md',
        model: { name: 'test-project-name' },
        project: { name: 'test-project-name', url: 'modules.md' },
        contents: 'CONTENTS',
      } as PageEvent;
      const { frontMatterComponent } = await generate();
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should set default index page with exports`, async () => {
      const page = {
        url: 'index.md',
        model: { name: 'test-project-name' },
        project: { name: 'test-project-name', url: 'index.md' },
        contents: 'CONTENTS',
      } as PageEvent;
      const { frontMatterComponent } = await generate({}, [
        '../typedoc-plugin-markdown/test/stubs/src/theme.ts',
      ]);
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should set custom readmeLabel and readmeTitle`, async () => {
      const page = {
        url: 'index.md',
        model: { name: 'test-project-name' },
        project: { name: 'test-project-name', url: 'modules.md' },
        contents: 'CONTENTS',
      } as PageEvent;
      const { frontMatterComponent } = await generate({
        sidebar: {
          readmeLabel: 'Custom readme label',
        },
        readmeTitle: 'Custom readme title',
      });
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should set the same custom readmeLabel and readmeTitle`, async () => {
      const page = {
        url: 'index.md',
        model: { name: 'test-project-name' },
        project: { name: 'test-project-name', url: 'modules.md' },
        contents: 'CONTENTS',
      } as PageEvent;
      const { frontMatterComponent } = await generate({
        sidebar: {
          readmeLabel: 'Custom readme title',
        },
        readmeTitle: 'Custom readme title',
      });
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });
  });

  describe(`(globals)`, () => {
    test(`should set default globals page`, async () => {
      const page = {
        url: 'modules.md',
        project: { name: 'test-project-name', url: 'modules.md' },
        contents: 'CONTENTS',
      } as PageEvent;
      const { frontMatterComponent } = await generate();
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should set custom indexLabel`, async () => {
      const page = {
        url: 'modules.md',
        project: { name: 'test-project-name', url: 'modules.md' },
        contents: 'CONTENTS',
      } as PageEvent;
      const { frontMatterComponent } = await generate({
        sidebar: {
          indexLabel: 'Custom index label',
        },
      });
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should set globals page when readme=none`, async () => {
      const page = {
        url: 'modules.md',
        project: { name: 'test-project-name', url: 'modules.md' },
        contents: 'CONTENTS',
      } as PageEvent;
      page.url = 'index.md';
      page.project.url = 'index.md';
      const { frontMatterComponent } = await generate();
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });
  });

  describe(`(reflection names)`, () => {
    test(`should return reflection labels with short names`, async () => {
      const { project, frontMatterComponent } = await generate();
      const page = {
        project,
        model: project.findReflectionByName('FrontMatterClass'),
        url: 'url',
        contents: 'CONTENTS',
      } as PageEvent;
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should return front matter with full names`, async () => {
      const { project, frontMatterComponent } = await generate({
        sidebar: {
          fullNames: true,
        },
      });

      const page = {
        project,
        model: project.findReflectionByName('FrontMatterClass'),
        url: 'url',
        contents: 'CONTENTS',
      } as PageEvent;
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });
  });
});
