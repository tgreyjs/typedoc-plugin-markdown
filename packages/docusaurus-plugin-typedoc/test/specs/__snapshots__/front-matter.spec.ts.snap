// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FrontMatter: (globals) should set custom indexLabel 1`] = `
"---
id: \\"modules\\"
title: \\"test-project-name\\"
sidebar_label: \\"Custom index label\\"
sidebar_position: 0.5
hide_table_of_contents: true
custom_edit_url: null
---

CONTENTS"
`;

exports[`FrontMatter: (globals) should set default globals page 1`] = `
"---
id: \\"modules\\"
title: \\"test-project-name\\"
sidebar_label: \\"Table of contents\\"
sidebar_position: 0.5
hide_table_of_contents: true
custom_edit_url: null
---

CONTENTS"
`;

exports[`FrontMatter: (globals) should set globals page when readme=none 1`] = `
"---
id: \\"index\\"
title: \\"test-project-name\\"
slug: \\"/api\\"
sidebar_label: \\"Table of contents\\"
sidebar_position: 0.5
hide_table_of_contents: true
custom_edit_url: null
---

CONTENTS"
`;

exports[`FrontMatter: (readme) should set custom readmeLabel and readmeTitle 1`] = `
"---
id: \\"index\\"
title: \\"Custom readme title\\"
slug: \\"/api\\"
sidebar_label: \\"Custom readme label\\"
sidebar_position: 0
custom_edit_url: null
---

CONTENTS"
`;

exports[`FrontMatter: (readme) should set default index page 1`] = `
"---
id: \\"index\\"
title: \\"test-project-name\\"
slug: \\"/api\\"
sidebar_label: \\"Readme\\"
sidebar_position: 0
custom_edit_url: null
---

CONTENTS"
`;

exports[`FrontMatter: (readme) should set default index page with exports 1`] = `
"---
id: \\"index\\"
title: \\"test-project-name\\"
slug: \\"/api\\"
sidebar_label: \\"Exports\\"
sidebar_position: 0.5
custom_edit_url: null
---

CONTENTS"
`;

exports[`FrontMatter: (readme) should set the same custom readmeLabel and readmeTitle 1`] = `
"---
id: \\"index\\"
title: \\"Custom readme title\\"
slug: \\"/api\\"
sidebar_position: 0
custom_edit_url: null
---

CONTENTS"
`;

exports[`FrontMatter: (reflection names) should return front matter with full names 1`] = `
"---
id: \\"url\\"
title: \\"Class: FrontMatterClass\\"
sidebar_label: \\"frontmatter.FrontMatterClass\\"
custom_edit_url: null
---

CONTENTS"
`;

exports[`FrontMatter: (reflection names) should return reflection labels with short names 1`] = `
"---
id: \\"url\\"
title: \\"Class: FrontMatterClass\\"
sidebar_label: \\"FrontMatterClass\\"
custom_edit_url: null
---

CONTENTS"
`;
